recipes post for version 0-1-14 (#477)
tidyverse/tidyverse.org: content/blog/recipes-0-1-14/index.Rmd

  )
theme_set(thm)
```

We're stoked to announce the release of [recipes](https://recipes.tidymodels.org) 0.1.14. recipes is an alternative method for creating and preprocessing design matrices that can be used for modeling or visualization. 

You can install it from CRAN with:

```{r, eval = FALSE}
install.packages("recipes")
```

You can see a full list of changes in the [release notes](https://recipes.tidymodels.org/news/index.html). There are some improvements and changes to talk about. 

## An alternative to `juice()`

Now that we have taught with recipes for a few years, we've realized that there is a lot of  confusion about the differences between `juice()` and `bake()`:

* `juice(recipe)` returns the preprocessed training set (at very low computational costs).
* `bake(recipe, new_data)` applies the recipe to any data (e.g. training, testing, unknowns, etc.)

We were not able to find ways to make this distinction clear for many users. 

How could we solve this issue? We decided to come up with a better alternative to `juice()`  that would be more intuitive. As a result,  _all_ applications of the recipe can now use `bake()`: 

> * `bake(recipe, new_data = some_data_set)` works as usual. 

Reviewer: ```suggestion
* `bake(recipe, new_data = some_data_set)` works as before. 
```

> * `bake(recipe, new_data = NULL)` returns the preprocessed training set. 

Reviewer: ```suggestion
* `bake(recipe, new_data = NULL)` now returns the preprocessed training set. 
```

This is precedented in base R since many `predict()` methods re-predict the training set when the `newdata` argument is `NULL` or missing. Note that there is no default for `new_data`; you have to set it to `NULL` to get the training set. 

We felt that this was the best API change that we could make. An external poll showed some agreement: 

```{r poll, echo  = FALSE, fig.align='center', out.width="80%"}
knitr::include_graphics("juice-poll.png")
```

> `juice()`, which is still my favorite R function name of all time, will not be removed; you can still use it. However, we will not use it in training materials or documentation. 

Reviewer: ```suggestion
`juice()`, which is still my favorite R function name of all time, will not be removed; you can still use it. However, we will not use it in training materials or most documentation. 
```
Since it is still a documented function

## Imputation with linear models

> Tim Zhou added a step to use linear models for imputation. This is a nice, compact method for embedding an imputation equation for numeric predictors into the recipe. The syntax is similar to the existing imputation steps. Here's an example from the Ames data: 

Reviewer: ```suggestion
Tim Zhou contributed a step to use linear models for imputation. This is a nice, compact method for adding an imputation equation for numeric predictors into the recipe. The syntax is similar to the existing imputation steps. Here's an example from the Ames data: 
```

```

We might be able to reasonably approximate the missing values based on the other geographic predictors (`Longitude` and `Neighborhood`) as well as a few aspects of the houses (e.g., `MS_Zoning` and `Alley`). A linear model is create with these predictors in order to estimate the missing `Latitude` data: 

```{r lm-imp}
imputed_ames <-
  recipe(Sale_Price ~ ., data = ames_missing) %>%
  step_impute_linear(
    Latitude,
    impute_with = imp_vars(Longitude, Neighborhood, MS_Zoning, Alley), 
    id = "lm-imp"
  ) %>%
  prep(ames_missing)
```

This plot shows the missing data's true values on the x-axis and their imputed values on the y-axis: 

```{r plot-values, echo = FALSE, fig.height=6, fig.width=6, out.width="70%", dev = 'svg', dev.args = list(bg = "transparent")}
imputed <-
  bake(imputed_ames, new_data = ames_missing) %>%
  dplyr::rename(imputed = Latitude) %>%
  bind_cols(ames %>% dplyr::select(original = Latitude)) %>%
  bind_cols(ames_missing %>% dplyr::select(Latitude)) %>%
  dplyr::filter(is.na(Latitude))

ggplot(imputed, aes(x = original, y = imputed)) +
  geom_abline(col = "green") +
  geom_point(alpha = .3) +
  coord_obs_pred()
```

> In future versions, we will standardize on the naming convention `step_impute_*()`. The existing functions will be soft-deprecated for a reasonable to ensure backward compatibility.  

Reviewer: ```suggestion
In future versions, we will standardize on the naming convention `step_impute_*()`. The existing functions will be soft-deprecated for a reasonable time period to ensure backward compatibility.  
```

## Other changes

> This version of the package has an extra option for `prep()` that will print some information on the differences in the data before and after the step was prepared: 

Reviewer: ```suggestion
This version of the package has an extra logging option for `prep()` that will print some information on the differences in the data before and after the step was prepared: 
```

```{r logging}
ames_rec <- recipe(Sale_Price ~ ., data = ames) %>%
  step_BoxCox(Lot_Area, Gr_Liv_Area) %>%
  step_other(Neighborhood, threshold = 0.05)  %>%
  step_dummy(all_nominal()) %>%
  step_interact(~ starts_with("Central_Air"):Year_Built) %>%
  step_ns(Longitude, Latitude, deg_free = 5) %>% 
  prep(log_changes = TRUE)
```

> Another important change was behind-the-scenes. There were problems with using PSOCK clusters on Windows because the worker processes were not aware of all the packages that should be loaded. recipes now ensures that all of the packages required by each step will be accessible in parallel. A similar change is coming to the parsnip package. 

Reviewer: ```suggestion
Another important change was behind the scenes. Before, there were problems with using PSOCK clusters on Windows because the worker processes were not aware of all the packages that should be loaded. Now, recipes ensures that all of the packages required by each step will be accessible in parallel. A similar change is coming soon to the parsnip package. 
```
